src/Intervals.php

```diff
@@ -100,11 +105,98 @@ public static function haveIntersections(ConstraintInterface $a, ConstraintInter
             return true;
         }
 
+        if ($a instanceof MatchNoneConstraint || $b instanceof MatchNoneConstraint) {
+            return false;
+        }
+
         $intersectionIntervals = self::generateIntervals(new MultiConstraint(array($a, $b), true), true);
 
         return \count($intersectionIntervals['numeric']) > 0 || \count($intersectionIntervals['branches']) > 0;
     }
 
+    /**
+     * Attempts to optimize a MultiConstraint
+     *
+     * When merging MultiConstraints together they can get very large, this will
+     * compact it by looking at the real intervals covered by all the constraints
+     * and then creates a new constraint containing only the smallest amount of rules
+     * to match the same intervals.
+     *
+     * @return ConstraintInterface
+     */
+    public static function compactConstraint(ConstraintInterface $constraint)
+    {
+        if (!$constraint instanceof MultiConstraint) {
+            return $constraint;
+        }
+
+        $intervals = self::generateIntervals($constraint);
+        $constraints = array();
+        $hasNumericMatchAll = false;
+        $isConjunctive = false;
+        for ($i = 0, $count = \count($intervals['numeric']); $i < $count; $i++) {
+            $interval = $intervals['numeric'][$i];
+            if ((string) $interval->getStart() === (string) Interval::zero() && (string) $interval->getEnd() === (string) Interval::positiveInfinity()) {
+                // if we matched != x then no need to add a constraint matching the whole numeric range
+                if ($isConjunctive) {
+                    break;
+                }
+                $constraints[] = $interval->getStart();
+                $hasNumericMatchAll = true;
+                break;
+            }
+
+            // convert back >= x - <= x intervals to == x
+            if ($interval->getStart()->getVersion() === $interval->getEnd()->getVersion() && $interval->getStart()->getOperator() === '>=' && $interval->getEnd()->getOperator() === '<=') {
+                $constraints[] = new Constraint('==', $interval->getStart()->getVersion());
+                continue;
+            }
+
+            // convert back 0 - <x + >x - +inf to != x
+            if (isset($intervals['numeric'][$i+1]) && $interval->getEnd()->getVersion() === $intervals['numeric'][$i+1]->getStart()->getVersion() && $interval->getEnd()->getOperator() === '<' && $intervals['numeric'][$i+1]->getStart()->getOperator() === '>') {
+                $constraints[] = new Constraint('!=', $interval->getEnd()->getVersion());
+                $intervals['numeric'][$i+1] = new Interval($interval->getStart(), $intervals['numeric'][$i+1]->getEnd());
+                $isConjunctive = true;
```

Comment:
Since you aren't doing this only if the intervals go from 0 and then to +inf, but for all of these types of intervals, I don't think you can just switch conjuntive here? What about:
  
`<3 || (>3, <5) || (>=6, <9)  => [0, 3) (3, 5) [6, 9)`

I think the result here would be an incorrect `!=3 && <5 && (>= 6, <9) => []`

Reply:
Ok fixed this by adding AnyDevConstraint and only optimizing to `!= x` when only those are present in the output. Definitely would need to add a few more tests to ensure correctness, but I think it's safe now, and can then be optimized further later if needed.